### ISSUE TITLE
feat: Support for Concurrent and DAG-based inputs and steps

### ISSUE THREAD

**Issue by mumoshu** [mumoshu/variant#125]
`Variant` has several shortcomings in terms of how it handles steps and inputs:

- Inputs take a long time to resolve when there are many, as it doesn't support concurrency
- Steps can't run concurrently
- An input can't depend on another "sibling" inputs
- Steps and Inputs are very similar to how it's executed and their results are referenced/reused

As of today, in below examples, steps and inputs are run sequentially and there's no way to make them concurrent:

```yaml
tasks:
  test:
    steps:
    - script: kubectl apply -f app1.yaml
    - script: helm upgrade --install charts/app2
```

```yaml
tasks:
  test:
    inputs:
    - name: install-app1
    - name: install-app2
```

I'd like to introduce this feature so that you can write a DAG of inputs and steps. Inputs and steps are merged into a same concept called Job. For instance:

```yaml
tasks:
  test:
    workflow:
      app1-enabled:
      app1:
        from:
        - env
        - config
      app1-dir:
      app1:
        task: {kubectl-apply: {file: "{{ .app1Dir }}"}}
        condition: .app1Enabled
        needs: [app1-enabled, app1-dir]
        #after: [parma1] if you don't refer to `.Deps.param1` from the template and only want to control execution order, not the existence of param1
      app2:
        task: {helm-upgrade: {chart: charts/mychart}}
```

Running this example with Variant would look like it has two concurrent "threads" each runs:

1. app1-enabled, then app1-dir, and finally app1
2. app2

Which results in concurrently installing two K8s apps with kubectl and helm respectively

Subsequent steps can reference results of preceeding steps like seen in `arg1: {{ .app1Dir }}` in the above example. kebab-cased names are lowerCamelCased to comply with go tmpl.

**Comment by mumoshu** [comment]
Closing this as covered by Variant 2. See #129
